In [1]:
import warnings

warnings.filterwarnings("ignore")
import numpy as np
import datasets
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split, ShuffleSplit, RandomizedSearchCV
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
from scipy.stats import loguniform

dni = 308  # Completar


ds = datasets.load_dataset("ylecun/mnist").with_format("numpy")
X_train = ds["train"][:]["image"].astype(np.float32).reshape(-1, 28 * 28) / 255.0
y_train = ds["train"][:]["label"].astype(np.uint8)
X_test = ds["test"][:]["image"].astype(np.float32).reshape(-1, 28 * 28) / 255.0
y_test = ds["test"][:]["label"].astype(np.uint8)

max_K = np.min(X_train.shape)
pca = PCA(n_components=max_K).fit(X_train)  # Completar

X_train = pca.transform(X_train)  # Completar
X_test = pca.transform(X_test)  # Completar

Ks = np.array([50, 100, 200, 300, 400, 500, max_K])  # Completar

clf = QuadraticDiscriminantAnalysis()
G = {"reg_param": loguniform(1e-9, 1e-1)}  # Completar
splitter = ShuffleSplit(n_splits=1, test_size=0.1, random_state=dni)
RS = RandomizedSearchCV(
    clf,
    G,
    n_iter=3,
    scoring="accuracy",
    n_jobs=-1,
    cv=splitter,
    pre_dispatch=32,
    random_state=dni,
)
for _, k in enumerate(reversed(np.sort(Ks))):
    Z_train = X_train[:, :k]  # Completar
    Z_test = X_test[:, :k]  # Completar

    acc = RS.fit(Z_train, y_train).score(Z_test, y_test)  # Completar
    print(f"Precisión: {acc:.2%} con {RS.best_params_} y K = {k}")



README.md: 0.00B [00:00, ?B/s]

mnist/train-00000-of-00001.parquet:   0%|          | 0.00/15.6M [00:00<?, ?B/s]

mnist/test-00000-of-00001.parquet:   0%|          | 0.00/2.60M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/60000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/10000 [00:00<?, ? examples/s]

Precisión: 92.26% con {'reg_param': np.float64(0.0018867019057725225)} y K = 784
Precisión: 92.71% con {'reg_param': np.float64(0.0018867019057725225)} y K = 500
Precisión: 93.69% con {'reg_param': np.float64(0.0018867019057725225)} y K = 400
Precisión: 94.17% con {'reg_param': np.float64(0.0018867019057725225)} y K = 300
Precisión: 95.07% con {'reg_param': np.float64(0.0018867019057725225)} y K = 200
Precisión: 95.92% con {'reg_param': np.float64(0.0018867019057725225)} y K = 100
Precisión: 96.35% con {'reg_param': np.float64(0.0018867019057725225)} y K = 50


A la vista de los resultados, qué combinacion de valores para `K` y `reg_param` elegirías? Por qué?

**Respuesta:**
Elegiría la combinación de reg_param = 0.0018867019057725225 y K = 100,
 ya que ofrecen la mejor precisión de todas las combinaciones (95.92%) al con una reducción de dimensión notable (784 -> 100)